### PR TITLE
CVSL-1505 Move licence creation to Backend

### PR DESCRIPTION
**Pull request by andrewrlee**
All the unit tests are complete.. still working on fixing the integration tests.
The create variation tests seem to be failing for some reason and not sure why?

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus.IN_PROGRESS
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus.REJECTED
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus.SUBMITTED
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceType
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceType.Companion.getLicenceType
+import java.time.LocalDateTime
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.entity.LicenceEvent as EntityLicenceEvent
+
+@Service
+class LicenceCreationService(
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt:35]: Extracted out createLicences into its own service.. all the code remains as is

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    if (licenceEqualityCheck.isNotEmpty()) {
+      log.warn("Licence inconsistency for ${licenceEntity.id} with the following differences: $licenceEqualityCheck")
+    } else {
+      log.info("Licence request and creation are identical")
+    }
+
+    return createLicenceResponse
+  }
+
+  private fun createLicenceRequest(
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt:140]: obviously we can merge the request and transforming logic once we are happy this works

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicencePolicyService.kt [hunk 3]
```diff
@@ -77,4 +79,32 @@ class LicencePolicyService(private val policies: List<LicencePolicy> = listOf(PO
...
+  fun getCurrentStandardPssConditions(licenceType: LicenceType): List<StandardCondition> {
+    return if (licenceType == LicenceType.AP) {
+      emptyList()
+    } else {
+      currentPolicy().standardConditions.standardConditionsPss.mapIndexed { index, standardConditionPss ->
+        StandardCondition(
+          code = standardConditionPss.code,
+          sequence = index,
+          text = standardConditionPss.text,
+        )
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicencePolicyService.kt:106]: once we're happy this works, we can have a transformer from the policy model to the entity

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PhoneDetail.kt [hunk 4]
```diff
@@ -4,5 +4,5 @@ data class PhoneDetail(
   val phoneId: Int,
   val number: String,
   val type: String,
-  val ext: String,
+  val ext: String?,
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PhoneDetail.kt:7]: probably need to check the nullability of these

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceCreationIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  @BeforeEach
+  fun reset() {
+    govUkApiMockServer.stubGetBankHolidaysForEnglandAndWales()
+  }
+
+  @Test
+  fun `Create a licence`() {
+    prisonApiMockServer.stubGetPrison()
+    prisonerSearchMockServer.stubSearchPrisonersByNomisIds()
+    probationSearchMockServer.stubSearchForPersonOnProbation()
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceCreationIntegrationTest.kt:46]: this is currently failing.. still need to set up all the stubbing

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PhoneDetail.kt (hunk 4 above)
Comment (gurps-singh) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PhoneDetail.kt:7]: I've been through and tried to match these - hopefully we will have a clearer picture once we test in dev!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt [hunk 6]
```diff
@@ -88,8 +89,12 @@ class LicenceServiceTest {
   private val notifyService = mock<NotifyService>()
   private val omuService = mock<OmuService>()
   private val releaseDateService = mock<ReleaseDateService>()
+  private val probationSearchApiClient = mock<ProbationSearchApiClient>()
```
Comment (gurps-singh) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt:92]: No longer need these four in the LicenceService as we now have the LicenceCreation service and these are only used in licence creation.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt [hunk 7]
```diff
@@ -124,6 +128,12 @@ class LicenceServiceTest {
       additionalConditionUploadDetailRepository,
       auditEventRepository,
       notifyService,
+      omuService,
+      releaseDateService,
+      probationSearchApiClient,
```
Comment (gurps-singh) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt:133]: These four can be removed as only used by licence creation which has been split to own service.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt (hunk 6 above)
Comment (gurps-singh) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt:92]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt (hunk 7 above)
Comment (gurps-singh) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceServiceTest.kt:133]: Done

**Review by gurps-singh — APPROVED** [review]

**Pull request merged**